### PR TITLE
Boss controller update

### PR DESCRIPTION
**Pull request by cdgonz115**
Just look at commit comments

### CODE REVIEW

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 1]
```diff
@@ -4,108 +4,156 @@
 using UnityEngine;
 using System.Linq;
 
-[RequireComponent (typeof(EnemyHealth))]
-public class BossBehaviorController : MonoBehaviour {
+public class BossAction : MonoBehaviour
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs]: Shouldnt this structure store a reference to an action?

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 2]
```diff
@@ -4,108 +4,156 @@
...
+        {
+            foreach (PhaseAction p in b.PossibleActions) p.behavior.enabled = false;
+            foreach (PhaseAction p in b.backgroundActions) p.behavior.enabled = false;
+            if (b.startBehaviour != null) b.startBehaviour.enabled = false;
+            if (b.endBehaviour != null) b.endBehaviour.enabled = false;
         }
         health = GetComponent<EnemyHealth>();
-        health.maxHealth = bossPhases.Select(x=> x.health).Sum();
-        health.onEnemyDamage+=checkHealth;
+        health.maxHealth = bossPhases.Select(x => x.health).Sum();
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs:46]: I think we should set up a linter for this project so that there aren't so many style changes in PRs
It makes it difficult to parse the actual changes in logic

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 3]
```diff
@@ -4,108 +4,156 @@
...
+    public IEnumerator StartPhase()
+    {
         phaseHealth = bossPhases[currentPhase].health;
-        if( bossPhases[currentPhase].startBehaviour != null) bossPhases[currentPhase].startBehaviour.enabled = true;
-        foreach(PhaseAction p in bossPhases[currentPhase].backgroundActions) StartCoroutine( startActionBackground(p) );
-        ChangeAction();
+        if (bossPhases[currentPhase].startBehaviour != null)
+        {
+            bossPhases[currentPhase].startBehaviour.enabled = true;
+            while (bossPhases[currentPhase].startBehaviour.actionRunning)
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs]: honestly i would have had a much more complex solution but this works
we can have higher order controllers without having to adjust this one too

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 4]
```diff
@@ -4,108 +4,156 @@
...
+
     }
-    public void ChangeAction () {
-        if(currentAction.behavior != null) currentAction.behavior.enabled = false;
-        if(actionQueue.Any()) { currentAction =  actionQueue[0]; actionQueue.RemoveAt(0); }
-        else if(bossPhases[currentPhase].PossibleActions.Count > 1) 
-            currentAction = bossPhases[currentPhase].PossibleActions.Where(x=> !x.Equals(currentAction)).ElementAt( rand.Next(bossPhases[currentPhase].PossibleActions.Count-1) );
+    public void ChangeAction()
+    {
+        foreach(PhaseAction p in actionQueue) print(p.behavior);
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs]: remove print statement when not needed, output is slow

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 5]
```diff
@@ -4,108 +4,156 @@
...
+        if (currentAction.behavior != null) currentAction.behavior.enabled = false;
+        if (bossPhases[currentPhase].repeatingBehaviour != null && currentAction.behavior != bossPhases[currentPhase].repeatingBehaviour) currentAction.behavior = bossPhases[currentPhase].repeatingBehaviour;
+        else if (actionQueue.Any()) { currentAction = actionQueue[0]; actionQueue.RemoveAt(0); }
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs]: this is an absolute mess, thats kinda my fault but 
god i cant parse it at all 
the diff makes it bad too
this needs to be spaced out more i dont know why i thought that should be on the same line

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 6]
```diff
@@ -4,108 +4,156 @@
...
+        else if (bossPhases[currentPhase].PossibleActions.Count > 1)
+            currentAction = bossPhases[currentPhase].PossibleActions.Where(x => !x.Equals(currentAction)).ElementAt(rand.Next(bossPhases[currentPhase].PossibleActions.Count));
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs]: Need to change this to evaluate with the change provided in the actions 
IE when the manager is started the numbers are statically evaluated to their percentages and replaced (so that the chance of all actions adds to 1) and then that number can be used probably, need to think about the math efficenty for selecting randomly from a list

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 7]
```diff
@@ -4,108 +4,156 @@
...
         else currentAction = bossPhases[currentPhase].PossibleActions.FirstOrDefault();
-
-        if(currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
+        if (currentAction.behavior != null) StartCoroutine(startAction(currentAction, true));
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs:131]: space these lines out from the current action selection logic beforhand

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 8]
```diff
@@ -4,108 +4,156 @@
...
+        public BossAction endBehaviour; // a behavior to enable on end
         public List<PhaseAction> backgroundActions; // behavours to be run on loop until the phase ends
-        public List<PhaseAction> PossibleActions; 
+        public List<PhaseAction> PossibleActions;
     }
     [Serializable]
-    public struct PhaseAction {
+    public struct PhaseAction
+    {
+        public bool independent;
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs:27]: what does this do?

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 9]
```diff
@@ -4,108 +4,156 @@
...
+    public IEnumerator startAction(PhaseAction action, bool startNewOnFinish)
+    {
         action.behavior.enabled = true;
+        action.behavior.actionRunning = true;
         int startedPhase = currentPhase;
-        for(int i = 0; i < action.runFrames; i++) {
-            if(currentPhase != startedPhase) {
-                action.behavior.enabled = false;
-                yield break;
+        if (action.independent || action.behavior == bossPhases[currentPhase].repeatingBehaviour)
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs:139]: oh is this for actions which have internal timers? 
that should probably be the default anyway

File: Assets/Scripts/Enemies/BossBehaviorController.cs [hunk 10]
```diff
@@ -4,108 +4,156 @@
...
+                if (currentPhase != startedPhase)
+                {
+                    action.behavior.enabled = false;
+                    yield break;
+                }
+                yield return new WaitForEndOfFrame();
+            }
+            action.behavior.enabled = false;
+        }
+        for (int i = 0; i < ((action.behavior == bossPhases[currentPhase].repeatingBehaviour) ? 0 : action.cooldownFrames); i++)
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/BossBehaviorController.cs:157]: why is this necessary

File: Assets/Scripts/Enemies/Ratboss/RatbossA0.cs [hunk 11]
```diff
@@ -4,10 +4,11 @@
 using System.Collections.Generic;
 using UnityEngine;
 
-public class RatbossA0 : MonoBehaviour
-{ 
+public class RatbossA0 : BossAction
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/Ratboss/RatbossA0.cs:7]: Okay so it does get inherited instead of MonoBehaviour
Please use `abstract` in the class definition to make that more clear

File: Assets/Scripts/Enemies/Ratboss/RatbossA3.cs [hunk 12]
```diff
@@ -88,34 +88,40 @@ void Awake()
 
     IEnumerator Racers()
     {
-        yield return A0.shakedoors();
-
         leftSpawn = Random.Range(yPosLR - (yLengthLR / 2), yPosLR + (yLengthLR / 2));
         racerL.transform.position = new Vector3(leftWall.transform.position.x + (xLengthLR / 2)+0.5f, leftSpawn, -2);
```
Comment (content-deleted) [inline Assets/Scripts/Enemies/Ratboss/RatbossA3.cs:92]: Oh god I never saw any of this code jesus fucking christ how horrifying

**Review by content-deleted — CHANGES_REQUESTED** [review]
If it works it works
There are a few things we can probably clean up here but I need to get the queue system actually working and make a few other changes

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 2 above)
Comment (content-deleted) [reply Assets/Scripts/Enemies/BossBehaviorController.cs:46]: I started doing a bit of research on this and there are options available 
I'm not gonna lie and say I have time to set it up right now but I could help with hosting if we wanted to added code quality gates and a jenkins pipeline to the project

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 4 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: oops pretty sure this was left because i was testing

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 8 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs:27]: since alot of the ratboss actions timings where handled by themselves in their own scripts i decided to add an option to the PhaseAction struct that lets the manager know if it should run the behavior either by the runframes variable which is handled by the bbc or the behavior themselves handle that, if the behavior themselves handle the timings then the behavior would be "independent"

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 9 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs:139]: yes lol

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 10 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs:157]: looking back i dont think it is, i think originally i had it out there but then decided to move it inside the if statement, still it be easiest to just put the action.behavior.enabled = false; after the if else statement

File: Assets/Scripts/Enemies/Ratboss/RatbossA0.cs (hunk 11 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/Ratboss/RatbossA0.cs:7]: will do

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 5 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: lol idk either

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 1 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: its an interface not a struct

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 3 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: yeah i thought this was the easiest way to handle a starting and  was planning to do the same for the ending action

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 5 above)
Comment (cdgonz115) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: i added some comments to help clear it up

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 1 above)
Comment (content-deleted) [reply Assets/Scripts/Enemies/BossBehaviorController.cs]: well its a class... not an interface (idk why i said structure instead of class because struct is an actual thing so saying (data) structure is confusing even if technically correct), interfaces cant describe what data is held in the classes which implement them, only give a guarantee of what methods exist on them etc

File: Assets/Scripts/Enemies/BossBehaviorController.cs (hunk 8 above)
Comment (nicholasngoc) [reply Assets/Scripts/Enemies/BossBehaviorController.cs:27]: Small suggestion to add a comment in here to describe this so all the other systems will for sure know what it is?

**Comment by danielwaker** [comment]
good for now
